File: src/main/java/io/github/kingg22/api/vacunas/panama/response/ApiResponseUtil.java
package io.github.kingg22.api.vacunas.panama.response;

import io.github.wimdeblauwe.errorhandlingspringbootstarter.ApiErrorResponse;
import io.github.wimdeblauwe.errorhandlingspringbootstarter.ApiFieldError;
import io.github.wimdeblauwe.errorhandlingspringbootstarter.ApiGlobalError;
import io.github.wimdeblauwe.errorhandlingspringbootstarter.ApiParameterError;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.constraints.NotNull;
import java.io.Serializable;
import java.time.Instant;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.context.request.ServletWebRequest;

/**
 * Utility class for handling and formatting API responses.
 * Provides helper methods to standardize metadata and create ResponseEntity objects.
 */
@Slf4j
public class ApiResponseUtil {
    private static final ApiResponseFactory apiResponseFactory = new ApiResponseFactory();

    private ApiResponseUtil() {
        throw new IllegalStateException("Utility class");
    }

    /**
     * Adds metadata information to the provided API response.
     * @param apiResponse {@link IApiResponse} object to which metadata will be added.
     * @param servletWebRequest {@link ServletWebRequest} request data used to extract information.
     */
    public static void setMetadata(
            @NotNull IApiResponse<?, Serializable> apiResponse, ServletWebRequest servletWebRequest) {
        log.debug("servletWebRequest: {}", servletWebRequest);
        log.debug("locale: {}", servletWebRequest.getLocale());
        apiResponse.addMetadata("path", servletWebRequest.getRequest().getRequestURI());
        apiResponse.addMetadata("timestamp", Instant.now().toString());
    }

    /**
     * Creates a standardized HTTP response from the API response.
     * @param apiResponse {@link IApiResponse} object containing the response with status.
     * @param webRequest {@link ServletWebRequest} used to set metadata in the response.
     * @return A {@link ResponseEntity} with the status code and body set to the API response object.
     */
    public static ResponseEntity<IApiResponse<String, Serializable>> sendResponse(
            IApiResponse<String, Serializable> apiResponse, ServletWebRequest webRequest) {
        setMetadata(apiResponse, webRequest);
        log.debug(apiResponse.toString());
        return ResponseEntity.status(apiResponse.getStatusCode()).body(apiResponse);
    }

    /**
     * Transforms an {@link ApiErrorResponse} from the error-handling library into a custom API response format.
     * Additionally, it hides internal errors in the response.
     * @param apiErrorResponse {@link ApiErrorResponse} object to be transformed.
     * @param request Additional request information, indicating the endpoint where the exception was thrown.
     * @return A new {@link IApiResponse} object.
     */
    public static IApiResponse<String, Serializable> transformApiErrorResponse(
            @NotNull ApiErrorResponse apiErrorResponse, Object request) {
        IApiResponse<String, Serializable> response = apiResponseFactory.createResponse();
        response.addStatus("code", apiErrorResponse.getHttpStatus().value());
        if (!apiErrorResponse.getMessage().contains("Dto")
                && !apiErrorResponse.getMessage().contains("api_vacunas_panama")) {
            response.addError(apiErrorResponse.getCode(), apiErrorResponse.getMessage());
        } else {
            response.addError(apiErrorResponse.getCode(), "Intente nuevamente");
        }
        for (ApiFieldError fieldError : apiErrorResponse.getFieldErrors()) {
            response.addError(fieldError.getCode(), fieldError.getProperty(), fieldError.getMessage());
        }
        for (ApiGlobalError globalError : apiErrorResponse.getGlobalErrors()) {
            response.addError(globalError.getCode(), globalError.getMessage());
        }
        for (ApiParameterError parameterError : apiErrorResponse.getParameterErrors()) {
            response.addError(parameterError.getCode(), parameterError.getParameter(), parameterError.getMessage());
        }
        switch (request) {
            case ServletWebRequest servletWebRequest -> setMetadata(response, servletWebRequest);
            case HttpServletRequest httpServletRequest -> setMetadata(
                    response, new ServletWebRequest(httpServletRequest));
            default -> response.addMetadata("timestamp", Instant.now().toString());
        }
        log.debug(response.toString());
        log.debug(
                "ErrorResponse(code: {}, message: {}, properties {})",
                apiErrorResponse.getCode(),
                apiErrorResponse.getMessage(),
                apiErrorResponse.getProperties());
        return response;
    }
}

File: src/main/kotlin/io/github/kingg22/api/vacunas/panama/response/ApiResponseUtil.kt
package io.github.kingg22.api.vacunas.panama.response

import io.github.kingg22.api.vacunas.panama.util.logger
import io.vertx.ext.web.RoutingContext
import jakarta.ws.rs.core.Response
import java.io.Serializable
import java.time.Instant

/**
 * Utility class for handling and formatting API responses. Provides helper methods to standardize metadata and create
 * ResponseEntity objects.
 */
object ApiResponseUtil {
    private val log = logger()

    /**
     * Adds metadata information to the provided API response.
     *
     * @param apiResponse [ApiResponse] object to which metadata will be added.
     * @param request [RoutingContext] request data used to extract information.
     */
    @JvmStatic
    fun setMetadata(apiResponse: ApiResponse, request: RoutingContext) {
        apiResponse.apply {
            addMetadata("path", request.request().path())
            addMetadata("timestamp", Instant.now().toString())
        }
    }

    /**
     * Adds metadata information to the provided API response.
     *
     * @param apiResponseBuilder [ApiResponseBuilder] object to which metadata will be added.
     * @param request [RoutingContext] request data used to extract information.
     */
    @JvmStatic
    fun setMetadata(apiResponseBuilder: ApiResponseBuilder, request: RoutingContext) {
        apiResponseBuilder.apply {
            withMetadata("path", request.request().path())
            withMetadata("timestamp", Instant.now().toString())
        }
    }

    /**
     * Creates a standardized HTTP response from the API response.
     *
     * @param apiResponse [ApiResponse] object containing the response with status.
     * @param request [RoutingContext] used to set metadata in the response.
     * @return A [Response] with the status code and body set to [ApiResponse] the API response object.
     */
    @JvmStatic
    fun createResponseEntity(apiResponse: ApiResponse, request: RoutingContext? = null): Response = apiResponse.apply {
        if (request != null) {
            setMetadata(this, request)
        }
        log.trace(toString())
    }.let {
        Response.status(it.retrieveStatusCode()).entity(it as ActualApiResponse).build()
    }

    /**
     * Creates a standardized HTTP response from the API response and build into ResponseEntity it immediately.
     *
     * @param request [RoutingContext] used to set metadata in the response.
     * @param data [Map] shortcut to add data to the response.
     * @param statusCode [Int] HTTP status code to be set in the response. Defaults to 200 OK.
     * @param builder [ApiResponseBuilder] builder to make the response.
     * @return A [Response] with the status code and body set to [ApiResponse] the API response object.
     */
    @JvmStatic
    fun createApiAndResponseEntity(
        request: RoutingContext,
        data: Map<String, Serializable>,
        statusCode: Int = 200,
        builder: ApiResponseBuilder.() -> Unit = {},
    ) = createResponseEntity(
        ApiResponseFactory.createResponseBuilder {
            apply(builder)
            withData(data)
            withStatusCode(statusCode)
        },
        request,
    )
    /*
        /**
     * Transforms an [ApiErrorResponse] from the error-handling library into a custom API response format.
     * Additionally, it hides internal errors in the response.
     *
     * @param apiErrorResponse [ApiErrorResponse] object to be transformed.
     * @param request Additional request information, indicating the endpoint where the exception was thrown.
     * @return A new [ApiResponse] object.
     */
        @JvmStatic
        fun transformApiErrorResponse(apiErrorResponse: ApiErrorResponse, request: ServerHttpRequest) =
            ApiResponseFactory.createResponseBuilder {
                withStatusCode(apiErrorResponse.httpStatus.value())

                val errorMessage = if (
                    apiErrorResponse.message.contains("Dto") ||
                    apiErrorResponse.message.contains("api.vacunas.panama")
                ) {
                    "Intente nuevamente"
                } else {
                    apiErrorResponse.message
                }

                withError(code = apiErrorResponse.code, property = null, message = errorMessage)

                apiErrorResponse.fieldErrors.forEach {
                    withError(code = it.code, message = it.message, property = it.property)
                }

                apiErrorResponse.globalErrors.forEach {
                    withError(code = it.code, message = it.message)
                }

                apiErrorResponse.parameterErrors.forEach {
                    withError(code = it.code, property = it.parameter, message = it.message)
                }

                setMetadata(this, request)

                log.trace(
                    "ErrorResponse(code: {}, message: {}, properties: {})",
                    apiErrorResponse.code,
                    apiErrorResponse.message,
                    apiErrorResponse.properties,
                )
            }
     */
}
